src/middleware/cookie_parser.rs

```diff
@@ -0,0 +1,159 @@
+        Self {
+            _parse: PhantomData,
+        }
+    }
+}
+
+impl<State, T> Middleware<State> for CookieParser<T>
+where
+    State: Send + Sync + 'static,
+    T: ParseCookies + Send + Sync,
+{
+    fn call(
+        &self,
+        mut request: Request<State>,
+        next: Next<State>,
+    ) -> BoxFuture<Result<Response, Error>> {
+        // Get the value of the `Cookie` header from the request.
+        let cookie_header = request.headers().get(COOKIE);
+
+        // Attempt to parse the value of the `Cookie` header if it exists and
+        // contains a valid cookie string.
+        //
+        let (parser_output, request_cookies) = match cookie_header.map(|value| value.to_str()) {
+            // There are no cookies to parse.
+            None => {
+                return next.call(request);
+            }
+
+            // The cookie header contains characters that are not visible ASCII.
+            Some(Err(error)) => {
+                let _ = error; // Placeholder for tracing...
+                return next.call(request);
+            }
+
+            // The cookie header contains a valid cookie string. Parse it.
+            Some(Ok(cookie_str)) => {
+                // Get an owned String from cookie_str.
+                let input = cookie_str.to_string();
+
+                // Parse the cookie string into an iter of results.
+                let mut output = T::parse_cookies(input).peekable();
+
+                // Get a mutable reference to the request cookies.
+                let cookies = if output.peek().is_some() {
+                    // The parser parsed some cookies. Allocate a new
+                    // `Box<CookieJar>` and get a mutable reference to it.
+                    request.cookies_mut()
+                } else {
+                    // The parser did not parse any cookies.
+                    return next.call(request);
+                };
+
+                (output, cookies)
+            }
+        };
+
+        // Iterate over the remaining results in the parser output. If a cookie
+        // was able to be parsed without error, add it to request_cookies.
+        //
+        parser_output.for_each(|result| match result {
+            Ok(cookie) => request_cookies.add_original(cookie),
+            Err(error) => {
+                let _ = error; // Placeholder for tracing...
+            }
+        });
+
+        // Clone request_cookies so we can merge them with the response cookies
+        // when they become available. This is necessary because we need to pass
+        // ownership of the request to the next middleware.
+        //
+        let mut merged_cookies = Box::new(request_cookies.clone());
+
+        Box::pin(async {
+            // Call the next middleware to get a response.
+            let mut response = next.call(request).await?;
+
+            if let Some(cookies) = response.cookies().map(CookieJar::iter) {
+                cookies.cloned().for_each(|cookie| {
```

Comment:
We can probably improve the performance of this middleware by using non `'static` cookies. However, when I think of cookies I think of sessions and authentication. These things require talking to external servers and / or databases which is often times the bottleneck anyways. I think this is a great solution in the short and long term.